src/main/java/picard/util/SequenceDictionaryUtils.java

```diff
@@ -0,0 +1,107 @@
+package picard.util;
+
+import htsjdk.samtools.SAMSequenceDictionaryCodec;
+import htsjdk.samtools.SAMSequenceRecord;
+import htsjdk.samtools.util.CloseableIterator;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.samtools.util.RuntimeIOException;
+import htsjdk.samtools.util.SortingCollection;
+import picard.PicardException;
+import picard.sam.CreateSequenceDictionary;
+
+import java.io.*;
+import java.util.Iterator;
+
+/**
+ * Created by farjoun on 1/12/18.
+ */
+public class SequenceDictionaryUtils {
+
+
+    /** Encodes a sequence dictionary
+     *
+     *
+     * @param writer a Buffered writer into which the dictinoary will be written
+     * @param samSequenceRecordIterator an iterator that produces SAMSequenceRecords
+     *
+     * @throws IllegalArgumentException if the iterator produces two SAMSequenceRecord with the same name
+     */
+    public static void encodeDictionary(final BufferedWriter writer, Iterator<SAMSequenceRecord> samSequenceRecordIterator) {
+        final SAMSequenceDictionaryCodec samDictCodec = new SAMSequenceDictionaryCodec(writer);
+
+        samDictCodec.encodeHeaderLine(false);
+        // SortingCollection is used to check uniqueness of sequence names
+        final SortingCollection<String> sequenceNames = makeSortingCollection();
+
+        // read reference sequence one by one and write its metadata
+        while (samSequenceRecordIterator.hasNext()) {
+            final SAMSequenceRecord samSequenceRecord = samSequenceRecordIterator.next();
+            samDictCodec.encodeSequenceRecord(samSequenceRecord);
+            sequenceNames.add(samSequenceRecord.getSequenceName());
+        }
+
+        // check uniqueness of sequences names
+        final CloseableIterator<String> iterator = sequenceNames.iterator();
+
+        if(!iterator.hasNext()) return;
+
+        String current = iterator.next();
+        while (iterator.hasNext()) {
+            final String next = iterator.next();
+            if (current.equals(next)) {
+                throw new PicardException("Sequence name " + current +
+                        " appears more than once in reference file");
+            }
+            current = next;
+        }
+    }
+
+    public static SortingCollection<String> makeSortingCollection() {
+        final File tmpDir = IOUtil.createTempDir("SamDictionaryNames", null);
+        tmpDir.deleteOnExit();
+        // 256 byte for one name, and 1/10 part of all memory for this, rough estimate
+        long maxNamesInRam = Runtime.getRuntime().maxMemory() / 256 / 10;
+        return SortingCollection.newInstance(
+                String.class,
+                new StringCodec(),
+                String::compareTo,
+                (int) Math.min(maxNamesInRam, Integer.MAX_VALUE),
+                tmpDir
+        );
+    }
+
+    private static class StringCodec implements SortingCollection.Codec<String> {
```

Comment:
I Agree....so it should be moved to htsjdk and then removed here...sharing it would require exposing a public class which would then require an api change to remove...

Reply:
also, I'm not sure I see where the code duplication is....

Reply:
My mistake, the other copies aren't in picard. One copy is in an htsjdk test, `SortingCollectionTest.StringCodec`, and the other is in picard-private `StringSortingCollectionFactory.StringCodec`.